src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -2288,74 +2288,32 @@ void PhaseIdealLoop::do_unroll(IdealLoopTree *loop, Node_List &old_new, bool adj
       new_limit = _igvn.intcon(limit->get_int() - stride_con);
       set_ctrl(new_limit, C->root());
     } else {
-      // Limit is not constant.
-      assert(loop_head->unrolled_count() != 1 || has_ctrl(opaq), "should have opaque for first unroll");
-      if ((stride_con > 0 && (java_subtract(limit_type->_lo, stride_con) < limit_type->_lo)) ||
-          (stride_con < 0 && (java_subtract(limit_type->_hi, stride_con) > limit_type->_hi))) {
-        // No underflow.
-        new_limit = new SubINode(limit, stride);
+      // Limit is not constant. Int subtraction could lead to underflow.
+      // (1) Convert to long.
+      Node* limit_l = new ConvI2LNode(limit);
+      register_new_node(limit_l, get_ctrl(limit));
+      Node* stride_l = new ConvI2LNode(stride);
+      register_new_node(stride_l, get_ctrl(limit));
```

Comment:
Can we make Long constant (`_igvn.longcon(stride)`) instead since `stride` is constant? Similar to `underflow_clamp_l`. My concern is you set control to constant which is not `Root`.

Reply:
Good point, will replace it with constant. Yes, I had the ctrl wrong, it should be root.